src/compiler/frontend/scanner/scanner.lisp

```diff
@@ -0,0 +1,178 @@
+(in-package :cl-braces.compiler.frontend.scanner)
+
+(deftype tpe-token-type ()
+  "The type of a token. This is a simple enumeration of all the different kinds of tokens that can be encountered in the"
+  '(member
+    :tok-illegal
+    :tok-eof))
+
+(defclass token ()
+  ((type :reader token-type :initarg :type :initform (error "no type given") :type tpe-token-type)
+   (text :reader token-text :initarg :text :initform (error "no text given") :type string)
+   (value :reader token-value :initarg :value :initform nil :type (or null t))
+   (location :reader token-location :initarg :location :initform (error "no location given") :type source-location))
+  (:documentation "A token is a single unit of input. It is the smallest unit of input that the parser can work with."))
+
+(defmethod print-object ((token token) stream)
+  (with-slots (type text value location) token
+    (print-unreadable-object (token stream :type t :identity t)
+      (format stream "type: ~a text: ~a value: ~a location: ~a" type text value location))))
+
+(defparameter *fail-fast* nil "If true, the scanner will enter the debugger when an error is encountered")
+
+(define-condition scan-error (error)
+  ((message :initarg :message :reader scan-error-message)
+   (location :initarg :location :reader scan-error-location))
+  (:report (lambda (condition stream)
+             (format stream "Illegal token at ~a" (scan-error-location condition)))))
+
+(defclass state ()
+  ((input :reader state-input :initarg :input :initform (error "no input given") :type source-input :documentation "The input that is being scanned. Must be a character stream.")
+   (input-stream :initform nil :type (or null stream) :documentation "The input stream that is being scanned. Must be a character stream.")
+
+   (consumed :initform (make-array 0 :element-type 'character :adjustable t :fill-pointer 0) :type (vector character) :documentation "The characters that have been consumed so far")
+   (token-offset  :initarg :start :initform 0 :type integer :documentation "The start position of the current token")
+   (token-column :initarg :column :initform 1 :type integer :documentation "The column in the input stream, where the current token starts")
+   (token-line  :initarg :line :initform 1 :type integer :documentation "The line in the input stream, where the current token starts")
+
+   (stream-line  :initarg :line :initform 1 :type integer :documentation "The line in the input stream, where the current token starts")
+   (stream-column :initarg :column :initform 1 :type integer :documentation "The column in the input stream, where the current token starts")
+   (stream-offset :initarg :offset :initform 0 :type integer :documentation "The offset in the input stream, where the current token starts"))
```

Comment:
Update the comments to be correct.
This is actually the Offset where the scanner currently points at